envbuilder.go

```diff
@@ -902,6 +838,340 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
+				opts.Logger(log.LevelError, "Falling back to the default image...")
+			}
+		}
+	} else {
+		// If a Dockerfile was specified, we use that.
+		dockerfilePath := filepath.Join(opts.WorkspaceFolder, opts.DockerfilePath)
+
+		// If the dockerfilePath is specified and deeper than the base of WorkspaceFolder AND the BuildContextPath is
+		// not defined, show a warning
+		dockerfileDir := filepath.Dir(dockerfilePath)
+		if dockerfileDir != filepath.Clean(opts.WorkspaceFolder) && opts.BuildContextPath == "" {
+			opts.Logger(log.LevelWarn, "given dockerfile %q is below %q and no custom build context has been defined", dockerfilePath, opts.WorkspaceFolder)
+			opts.Logger(log.LevelWarn, "\t-> set BUILD_CONTEXT_PATH to %q to fix", dockerfileDir)
+		}
+
+		dockerfile, err := opts.Filesystem.Open(dockerfilePath)
+		if err == nil {
+			content, err := io.ReadAll(dockerfile)
+			if err != nil {
+				return nil, fmt.Errorf("read Dockerfile: %w", err)
+			}
+			buildParams = &devcontainer.Compiled{
+				DockerfilePath:    dockerfilePath,
+				DockerfileContent: string(content),
+				BuildContext:      filepath.Join(opts.WorkspaceFolder, opts.BuildContextPath),
+			}
+		}
+	}
+
+	// When probing the build cache, there is no fallback!
+	if buildParams == nil {
+		return nil, fmt.Errorf("no Dockerfile or devcontainer.json found")
+	}
+
+	HijackLogrus(func(entry *logrus.Entry) {
+		for _, line := range strings.Split(entry.Message, "\r") {
+			opts.Logger(log.LevelInfo, "#%d: %s", stageNumber, color.HiBlackString(line))
+		}
+	})
+
+	if opts.LayerCacheDir != "" {
+		if opts.CacheRepo != "" {
+			opts.Logger(log.LevelWarn, "Overriding cache repo with local registry...")
+		}
+		localRegistry, closeLocalRegistry, err := serveLocalRegistry(ctx, opts.Logger, opts.LayerCacheDir)
+		if err != nil {
+			return nil, err
+		}
+		defer closeLocalRegistry()
+		opts.CacheRepo = localRegistry
+	}
+
+	// IgnorePaths in the Kaniko opts doesn't properly ignore paths.
+	// So we add them to the default ignore list. See:
+	// https://github.com/GoogleContainerTools/kaniko/blob/63be4990ca5a60bdf06ddc4d10aa4eca0c0bc714/cmd/executor/cmd/root.go#L136
+	ignorePaths := append([]string{
+		constants.MagicDir,
+		opts.WorkspaceFolder,
+		// See: https://github.com/coder/envbuilder/issues/37
+		"/etc/resolv.conf",
+	}, opts.IgnorePaths...)
+
+	if opts.LayerCacheDir != "" {
+		ignorePaths = append(ignorePaths, opts.LayerCacheDir)
+	}
+
+	for _, ignorePath := range ignorePaths {
+		util.AddToDefaultIgnoreList(util.IgnoreListEntry{
+			Path:            ignorePath,
+			PrefixMatchOnly: false,
+			AllowedPaths:    nil,
+		})
+	}
+
+	stdoutReader, stdoutWriter := io.Pipe()
+	stderrReader, stderrWriter := io.Pipe()
+	defer stdoutReader.Close()
+	defer stdoutWriter.Close()
+	defer stderrReader.Close()
+	defer stderrWriter.Close()
+	go func() {
+		scanner := bufio.NewScanner(stdoutReader)
+		for scanner.Scan() {
```

Comment:
Should these goroutines exit when the context is done?

Reply:
good point; we didn't need to worry about this before because the binary would just exit